### PR TITLE
icepeak: init at 0.7.2

### PR DESCRIPTION
**Pull request by rkrzr**
Icepeak is a fast JSON document store with push notification support.

###### Motivation for this change

Icepeak is one of our open-source projects and we would like to get it added to nixpkgs so that other people can install it with nix, instead of having to compile from source.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/icepeak/default.nix
Comment (jonringer) [inline pkgs/servers/icepeak/default.nix]: this doesn't work
```
cannot import '/nix/store/p4gzy8wbc92cmj0q1r10ksaws9w9nh4f-source/server', since path '/nix/store/mw17npnlqsmfbwbcm512km0wck900b5w-source.drv' is not valid, at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/hetzner-sb65-1083082-hel1-dc2/lib/customisation.nix:119:45
```
Also, having a default.nix in a repo is fine, but the responsibility for knowing how to build a package in nixpkgs should be nixpkgs, not deferred to the package. (personal opinion)

@Infinisil  opionion?

Comment (rkrzr) [reply pkgs/servers/icepeak/default.nix]: I understand the point of nixpkgs being the source of truth here. It is too bad that this doesn't work though, since that means that I'll have to copy the contents from the `default.nix` in the Icepeak repo to this file here (and repeat that process whenever we change the `default.nix` in the Icepeak repo).

I assumed that there would be some way to do this, since e.g. for packages that contain a Makefile you essentially also delegate the actual buildsteps to that external file.

Comment (jonringer) [reply pkgs/servers/icepeak/default.nix]: correct, but the main difference is that make is dumb, it will just make assumptions about the commands that can be executed, and as part of our packaging, we just fulfill those assumptions.

delegating the construction of the nix expression to the package creates the scenario where the package is making assumptions about what's contained in nixpkgs, and nixpkgs is making some assumptions about what it needs to provide for the package. Which blurs the lines of responsibility.

Furthermore, nixpkgs has many treewide changes, and if we wanted to do something like fix some flags or metadata, then we would need to open a PR against your repo, then update the src in nixpkgs, which doesn't seem right.

Comment (rkrzr) [reply pkgs/servers/icepeak/default.nix]: Ah okay, that makes sense. I'll update the PR to not depend on an external Nix file. Thanks for the explanation!

File: maintainers/maintainer-list.nix
Comment (jonringer) [inline maintainers/maintainer-list.nix]: the commit for this should be:
```
maintainers: add rkrzr
```

**Review by jonringer — COMMENTED** [review]
structure of this is a little non-standard, but I'm also not super familiar with haskell packaging

**Review by cdepillabout — CHANGES_REQUESTED** [review]
Could you try to get upstream to release this to hackage?

If they do so, then this will be automatically available in nixpkgs and we will hopefully not have to do any work to keep it updated.

--------------------

If upstream doesn't want to upload to hackage, then there are a couple things that should be changed here.

The biggest thing is that this file should be generated automatically with `cabal2nix`.  Take a look at how `spago` is packaged, and try to replicate that.

File: maintainers/maintainer-list.nix
Comment (rkrzr) [reply maintainers/maintainer-list.nix]: Okay, fixed.

**Comment by Infinisil** [comment]
To keep the Nix file in line with the one in the repo you could write a `passthrough.updateScript` that automatically downloads the file from the repo, see https://nixos.org/nixpkgs/manual/#ssec-stdenv-attributes for more info

**Comment by rkrzr** [comment]
> To keep the Nix file in line with the one in the repo you could write a passthrough.updateScript 

Thanks, I'll give this a try.

I also started the process to upload the package to Hackage (currently waiting for approval). Would that still be preferable?

**Comment by cdepillabout** [comment]
> I also started the process to upload the package to Hackage (currently waiting for approval). Would that still be preferable?

Yes, in general it is preferable to just get Haskell packages from Hackage (since the updating process is automatic).

**Comment by rkrzr** [comment]
> Yes, in general it is preferable to just get Haskell packages from Hackage (since the updating process is automatic).

Okay, I'll move forward with that then. Out of curiosity: How does the Hackage <-> Nixpkgs synchronizing process work? Does it simply pick the latest version on Hackage? Or does it also make use of Stackage snapshots somehow?

**Comment by cdepillabout** [comment]
@rkrzr I believe I wrote about this in the following post:

https://discourse.nixos.org/t/haskellpackages-stm-containers-fails-to-build/5416/4

If not, then you should probably be able to find an explanation about it in one of my past posts here or on discourse.

**Comment by rkrzr** [comment]
I added the package to Hackage now: https://hackage.haskell.org/package/icepeak-0.7.1.0

Is the process now to wait until `hackage2nix` is run again now, to pick up the new package? The [build](https://hydra.nixos.org/build/112872548#tabs-summary) seems to have been successful already.

**Comment by jonringer** [comment]
@peti runs it once a week, on thursday I believe.

**Comment by cdepillabout** [comment]
@rkrzr You should be able to find `icepeak` on the haskell-updates branch:

https://github.com/NixOS/nixpkgs/pull/79819

The `haskell-updates` branch gets merged into `master` about once a week by peti.  However, it didn't happen last week because of some hydra breakage.  I imagine he'll probably do it in a week or so.

Until that happens, I'd recommend doing the following (although you don't necessarily have to do any of these):

- Rebase this PR on the `haskell-updates` branch.
- Use `haskell.lib.justStaticExecutables` on icepeak.  (You can grep through the code to see some examples of people doing this.)
- Add yourself to the maintainers section for icepeak.  You can do this by adding yourself to this list: https://github.com/NixOS/nixpkgs/blob/5af23cdaf6be13b6de2c5ffc9d9b8ef83a3578f7/pkgs/development/haskell-modules/configuration-hackage2nix.yaml#L2569

**Comment by rkrzr** [comment]
OK, I have rebased this PR on `haskell-updates` now and applied the other two suggestions as well.

**Comment by jonringer** [comment]
you will probably want to rebase again...

**Comment by cdepillabout** [comment]
@rkrzr 

There seem to be two remaining problems here.

First, `haskell-updates` was just merged into `master` (and then rebased on top of `master`), so you'll need to rebase this one more time.

Second, The haskell stuff in nixpkgs was recently updated to LTS-15 (which contains GHC-8.8).

It looks like one of the dependencies of `icepeak`, `sqlite-simple-0.4.17.0`, does not compile with GHC-8.8:

```console
$ nix-build -A icepeak --arg config '{allowBroken=true;}'
these derivations will be built:
  /nix/store/6ax4v3rzafkh1dbvkxpilgsmkzyfkssl-sqlite-simple-0.4.17.0.drv
  /nix/store/skpiqb282jlsiypjcrqga5l5aqhxp77d-icepeak-0.7.1.0.drv
...
building '/nix/store/6ax4v3rzafkh1dbvkxpilgsmkzyfkssl-sqlite-simple-0.4.17.0.drv'...
...
Configuring sqlite-simple-0.4.17.0...
CallStack (from HasCallStack):
  die', called at libraries/Cabal/Cabal/Distribution/Simple/Configure.hs:1022:20 in Cabal-3.0.1.0:Distribution.Simple.Configure
  configureFinalizedPackage, called at libraries/Cabal/Cabal/Distribution/Simple/Configure.hs:475:12 in Cabal-3.0.1.0:Distribution.Simple.Configure
  configure, called at libraries/Cabal/Cabal/Distribution/Simple.hs:625:20 in Cabal-3.0.1.0:Distribution.Simple
  confHook, called at libraries/Cabal/Cabal/Distribution/Simple/UserHooks.hs:65:5 in Cabal-3.0.1.0:Distribution.Simple.UserHooks
  configureAction, called at libraries/Cabal/Cabal/Distribution/Simple.hs:180:19 in Cabal-3.0.1.0:Distribution.Simple
  defaultMainHelper, called at libraries/Cabal/Cabal/Distribution/Simple.hs:116:27 in Cabal-3.0.1.0:Distribution.Simple
  defaultMain, called at Setup.hs:5:8 in main:Main
Setup: Encountered missing or private dependencies:
semigroups ==0.18.*

builder for '/nix/store/6ax4v3rzafkh1dbvkxpilgsmkzyfkssl-sqlite-simple-0.4.17.0.drv' failed with exit code 1
cannot build derivation '/nix/store/skpiqb282jlsiypjcrqga5l5aqhxp77d-icepeak-0.7.1.0.drv': 1 dependencies couldn't be built
```

In order to work around this, I tried to jailbreak `sqlite-simple`:

```diff
diff --git a/pkgs/development/haskell-modules/configuration-common.nix b/pkgs/development/haskell-modules/configuration-common.nix
index 3f28ebebd22..26189384731 100644
--- a/pkgs/development/haskell-modules/configuration-common.nix
+++ b/pkgs/development/haskell-modules/configuration-common.nix
@@ -1467,4 +1467,8 @@ self: super: {
   # haskell-ci-0.8 needs cabal-install-parsers ==0.1, but we have 0.2.
   haskell-ci = doJailbreak super.haskell-ci;
 
+  # sqlite-simple-0.4.17.0 requires semigroups ==0.18.* while LTS-15 contains
+  # semigroups-0.19.
+  sqlite-simple = doJailbreak super.sqlite-simple;
+
 } // import ./configuration-tensorflow.nix {inherit pkgs haskellLib;} self super
```

However, there were still problems with `sqlite-simple`:

```console
$ nix-build -A icepeak --arg config '{allowBroken=true;}'
these derivations will be built:
  /nix/store/6ax4v3rzafkh1dbvkxpilgsmkzyfkssl-sqlite-simple-0.4.17.0.drv
  /nix/store/skpiqb282jlsiypjcrqga5l5aqhxp77d-icepeak-0.7.1.0.drv
...
building '/nix/store/6ax4v3rzafkh1dbvkxpilgsmkzyfkssl-sqlite-simple-0.4.17.0.drv'...
...
Building library for sqlite-simple-0.4.17.0..
[ 1 of 12] Compiling Database.SQLite.Simple.Ok ( Database/SQLite/Simple/Ok.hs, dist/build/Database/SQLite/Simple/Ok.o )

Database/SQLite/Simple/Ok.hs:75:5: error:
    ‘fail’ is not a (visible) method of class ‘Monad’
   |
75 |     fail str = Errors [SomeException (ErrorCall str)]
   |     ^^^^
...
builder for '/nix/store/mz1pcg808mdj1cmbx7vja7a1ggjgkmlx-sqlite-simple-0.4.17.0.drv' failed with exit code 1
cannot build derivation '/nix/store/zkqpgadm5blv9r5qprx68lxgllzl54d6-icepeak-0.7.1.0.drv': 1 dependencies couldn't be built
```

The easiest fix here would be to get the maintainer to make a release of `sqlite-simple` that supported the latest version of `semigroups` and GHC-8.8.  Then `icepeak` could just use that.

Also, if you want to get `sqlite-simple` and `icepeak` in Stackage, you won't have to worry about this problem in the future.

---------------------------------------

Sorry for going back and forth on this so much.

**Comment by imalsogreg** [comment]
FYI, the upstream issue in sqlite-simple is being discussed here: https://github.com/nurpax/sqlite-simple/issues/84

**Comment by rkrzr** [comment]
Alright, I now rebased this one more time on `haskell-updates` and also released a new version of Icepeak on Hackage which builds with GHC 8.8.

**Comment by cdepillabout** [comment]
@rkrzr This is looking good, but you'll need to do one more thing.

Please mark `icepeak` (and its dependency `sqlite-simple`) as non-broken.

Here's a video showing how to do this:

https://discourse.nixos.org/t/video-tutorial-how-to-fix-broken-haskell-packages-in-nix/3968

**Comment by rkrzr** [comment]
Ok, I've rebased this one more time and marked `icepeak` and `sqlite-simple` as non-broken now.

**Comment by cdepillabout** [comment]
@rkrzr Looks like everything builds now, thanks for following up on this!

**Pull request merged**
